Better logging if writing a tag/catetory/author page fails

MinchinWeb opened the pull request:
Add a short logging message if for some reason an invalid tag is provided, and Pelican instead tries to overwrite the index page.

Before:

```
           CRITICAL RuntimeError: File D:/Code/obsidian-vault/zz_output/tags/index.html is to be         __init__.py:666
                    overwritten
```

After:

```
[21:56:21] ERROR    Trying to write tag page for "**".                                                 generators.py:580
           CRITICAL RuntimeError: File D:/Code/obsidian-vault/zz_output/tags/index.html is to be         __init__.py:666
                    overwritten
```

(For some reason, the tag `**` is mapped onto the index page. Perhaps that is another issue that should be fixed...)

Fix is applied when writing Tag, Category, and Author pages. Let me know if it should be added other places.

# Pull Request Checklist

- [ ] Ensured **tests pass** and (if applicable) updated functional test output
- [ ] Conformed to **code style guidelines** by running appropriate linting tools
- [ ] Added **tests** for changed code
- [-] Updated **documentation** for changed code

avaris commented on pelican/generators.py:
```diff
@@ -560,57 +560,69 @@ def generate_tags(self, write):
+                    page_name=tag.page_name,
+                    all_articles=self.articles,
+                )
+            except RuntimeError as e:
+                logger.error('Trying to write Tag page for "%s".' % (tag))
+                raise e
```
* It's not "trying" anymore. It tried and now failed, maybe `"Unable to write..."` instead
* Please don't format logs in place (see: [docs](https://docs.getpelican.com/en/latest/contribute.html#logging-tips)): `logger.error("...", tag)`
* You don't do anything with exception (`as e`), so same as:
```python
try:
    # ...
except RuntimeError:
    # ...
    raise
```

Same for the others.

avaris commented:
> For some reason, the tag ** is mapped onto the index page.

Because, slug is empty and depending on `TAG_URL/TAG_SAVE_AS` you'll get "weird" results:

```python
>>> from pelican.urlwrappers import Tag
>>> from pelican.settings import DEFAULT_CONFIG
>>> t = Tag('**', DEFAULT_CONFIG)
>>> t.slug
''
```

And that's because `slugify` will strip/replace "questionable" characters. Since this will become part of the filename in the output, and filesystems can be picky.

Off the top of my head, I cannot think of any reason for wanting an empty slug intentionally. So if no objections, we could error for empty slug.

MinchinWeb commented:
I've updated it as per your comments! This is the current message:

```
[20:13:46] WARNING  Tag "**" has an invalid slug; skipping writing tag page...                                 log.py:89
```

I have it throw a warning (rather than a critical error) if the slug is invalid, but throws the "regular" critical error if it's run into some other problem.

I've "un- pre- formatted" the log messages, and dropped the `as e`.

Overall, it makes sense that slugify would strip out questionable character; I was just annoyed that Pelican would break but not really tell me what was wrong or how to fix it; I hope this explains how to fix the issue to the next user.

MinchinWeb commented on pelican/generators.py:
```diff
@@ -560,57 +560,90 @@ def generate_tags(self, write):
+                    all_articles=self.articles,
+                )
+            except RuntimeError:
+                if not tag.slug:
+                    logger.warning(
+                        'Tag "%s" has an invalid slug; skipping writing tag page...',
+                        tag,
+                    )
+                    continue
```
So during Tag creation?

This issue came up for me while working on a replacement Markdown reader, so I wouldn't want it to be tied directly to the reader, but when tags are created could work.

avaris replied:
Yes, e.g. here at the end: https://github.com/getpelican/pelican/blob/2d75ca8391bccb79bdac3fc7fd609a3a337e6fd8/pelican/urlwrappers.py#L13-L17

Something like:
```python
if not self.slug:
    logger.warning("%s `%s` will result in empty slug", self.__class__.__name__, self._name)
```

PS: Changed my mind and lowered it to a warning. Probably doesn't need to be an error.

MinchinWeb commented:
I've added a warning when trying to generate a slug from a name that won't give a valid slug, and downgraded the former warning to an INFO message. So the displayed messages (when run with `--verbose`) now look like:

```
           WARNING  Unable to generate valid slug for Tag "#**".                                               log.py:89
[...]
           INFO     Tag "#**" has an invalid slug; skipping writing tag page...                        generators.py:579
```

avaris commented on pelican/generators.py:
```diff
@@ -560,57 +560,93 @@ def generate_tags(self, write):
+                    all_articles=self.articles,
+                )
+            except RuntimeError:
+                if not tag.slug:
+                    logger.info(
+                        'Tag "%s" has an invalid slug; skipping writing tag page...',
+                        tag,
+                        extra={"limit_msg": "Further tags with invalid slugs."},
+                    )
+                    continue
```
With empty slugs already generating warnings, is this special case necessary anymore? Why not just do the `else` block?

MinchinWeb replied:
I think it's helpful to still keep this, because it makes it very explicity what Pelican is (or isn't doing), and it's not entirely obvious that an invalid slug would keep the page from being generated.

Also, it mirrors the logging for all the pages that *are* written.

MinchinWeb commented:
Hi @avaris !

This seems to have been lost in the shuffle.

Do you want me to remove the logging (of the non-generated pages), or is it good to merge as is?

Thanks!

MinchinWeb commented:
Hey @avaris ! Let me know if anything more is needed to merge this! Thanks!

avaris commented:
Apologies, offline life get in the way of things. I'll try to get back to this ASAP.

MinchinWeb commented:
> Apologies, offline life get in the way of things. I'll try to get back to this ASAP.

No worries at. My offline life has kept me from anything Pelican-related for the last couple months too...

minchinweb commented:
Hey @avaris , just wanted to check in on this. I'm maintaining a local fork with this on my side, but would love to just see it merged in.... Thanks for your help.

minchinweb commented:
Just wanted to bump this, hoping to get it merged! It's one of the patches I'm maintaining locally, and I was going to get up a new pull request till I remembered this is still here waiting...

Let me know if changes are needed; I'm happy to make them.

minchinweb commented on pelican/generators.py:
```diff
@@ -560,57 +560,93 @@ def generate_tags(self, write):
+                        tag,
+                        extra={"limit_msg": "Further tags with invalid slugs."},
+                    )
+                    continue
```
Done! (Though I called the exception `FileOverwriteFailedError` because I thought that would be clearer)

minchinweb commented:
(rebased to resolve merge conflicts)

justinmayer commented on pelican/writers.py:
```diff
@@ -107,14 +111,16 @@ def _open_w(self, filename, encoding, override=False):
-                raise RuntimeError(f"File {filename} is set to be overridden twice")
+                raise FileOverwriteFailedError(
+                    f"File {filename} is set to be overridden twice"
+                )
```
I have never fully understood the language in this error output. I know you weren't the person who authored this language, but could you take a moment and rewrite this error message such that it can be better understood by end users? (Including using "overwritten" instead of what I believe to be an incorrect usage here of "overridden".)

minchinweb replied:
I'll take a crack at this in the morning!

justinmayer commented on pelican/writers.py:
```diff
@@ -107,14 +111,16 @@ def _open_w(self, filename, encoding, override=False):
                 logger.info("Overwriting %s", filename)
             else:
-                raise RuntimeError(f"File {filename} is to be overwritten")
+                raise FileOverwriteFailedError(f"File {filename} is to be overwritten")
```
Same idea here: Could you take a moment and rewrite this error message such that it can be better understood by end users?

justinmayer reviewed (commented):
Hey @minchinweb. At the risk of prolonging this further, while you are here and editing these particular lines, could I possibly trouble you to rewrite two of the error messages, with full awareness that you were not the person who originally wrote them? I'd really appreciate it! 😊

minchinweb commented:
@justinmayer Error messages rewritten! (Overwritten?)
